### ISSUE TITLE
Boolean indexing/masking with the wrong shape should fail

### ISSUE THREAD

**Issue by shoyer** [google/jax#7329]
This example results in an error with NumPy but not jax.numpy:
```python
import jax.numpy as jnp

x = jnp.arange(4)
y = jnp.array([True, False])
x[y]   # DeviceArray([0], dtype=int32)
```

It looks like JAX converts the boolean indexer array into an integer array for indexing without checking the shape.

We should raise an error in this case: indexing an array of shape (4,) with a boolean array of shape (2,) is never valid.
